### DISCUSSION TITLE
Are nint/nuint intended to be used as a replacement for IntPtr/UIntPtr

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/csharplang#4368] [General]
As they technically do the same thing at a base level, are C#9+ codebases intended to use `nint` or `nuint` to replace all `IntPtr` or `UIntPtr` usages?

**Comment by huoyaoyuan** [comment]
> As they technically do the same thing at a base level

Yes. `nint` compiles into plain `IntPtr`.



> are C#9+ codebases intended to use `nint` or `nuint` to replace all `IntPtr` or `UIntPtr` usages?

I'm not sure. Do use `nint` when you need arithmetic operations.

**Comment by qrli** [comment]
No. `nint` and `nuint` are integers, while `IntPtr` and `UIntPtr` are pointers. It depends on what your intention is.

**Reply by tannergooding** [reply]
While you likely want to use `nint`/`nuint` when dealing with integral values and `IntPtr`/`UIntPtr` when dealing with opaque handles, having `Ptr` in the name has little to do with them being pointers.
The names come from C/C++ where `intptr_t` and `uintptr_t` exist as typedefs. They are named such because they are `ptr-bits` in length. Same as `int32_t` (`System.Int32`) is `32-bits` in length and `int64_t` (`System.Int64`) is `64-bits` in length.

`IntPtr`/`UIntPtr`, prior to the introduction of `nint`/`nuint`, were also less featured than actual pointers in terms of operators supported and so they might not even be desirable to use in all scenarios.
They continue to work well for opaque handles where you don't want to expose unsafe code as part of your public surface area. While `nint`/`nuint` work well whenever you have a proper integer, like a size, or alignment, or when you need to do certain arithmetic on a pointer (such as checking its alignment).

**Reply by qrli** [reply]
I agree that functionally `nint` can replace `IntPtr`. But in declaration/contract, they kind of convey different messages to developers. `IntPtr` is mostly used for pointers and handles, where consumers are typically not expected to manipulate the values. Surely it is only typing difference, just like `int` vs `void*` in C.

So, is the current recommendation to forget that typing? I can understand it may be well true in some frameworks, as they have already been working that way. But are you going to make it the general choice by default?

**Reply by Unknown6656** [reply]
I use `nint` or `nuint` for integral/arithmetic operations and `void*` for pointer operations. I nowadays hardly use `IntPtr` or `UIntPtr` at all.....

**Reply by ghost** [reply]
Okay so from my understanding here, in the context of Win32 something like `PVOID` should be an `IntPtr`/`UIntPtr` and something like `SIZE_T` should be an `nint`/`nuint`?

**Reply by qrli** [reply]
@Unknown6656 For applications, you can absolute do that. But for library authors, we also need to consider .NET languages other than C#, which may not support pointer types at all. That's one of the reasons `IntPtr` is a thing.

If .NET could say pointer type should be supported by all .NET languages, that would remove the need of `IntPtr`.

**Reply by huoyaoyuan** [reply]
Pointer is something you can dereference. `IntPtr`, or handle, is more opaque.
